Allow users to configure autopause settings for clusters

akarukappadath opened the pull request:
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

jmthibault79 committed:
Build script updates
- DB name changed to leotestdb
- Generate error when no arguments are supplied
- Remove old jars

jacmrob committed:
fix mysql build issue

jacmrob committed:
fix

jacmrob committed:
fix container link

jacmrob committed:
oops

jacmrob committed:
put it back

jacmrob committed:
another oops

jacmrob committed:
oops 3

helgridly committed:
GET cluster details endpoint that users can poll to see if they're ready yet (#20)

jacmrob committed:
change mysql port #

helgridly committed:
make swagger know about auth (#26)

helgridly committed:
Fix swagger auth (#27)

vkumra-broad committed:
GAWB-2271 - Delete cluster endpoint for leonardo (#22)

helgridly committed:
GIT_MODEL_HASH is the wrong word (#30)

rtitle committed:
[GAWB-2263] Notebook Service: Implement the GET /clusters endpoint for connecting to a notebook (#21)

* SSL proxy

* Use wss scheme

* Fix tests

* Minor cleanup

* Simplify test code

* Made ClusterDnsCache an Actor

* Fixes based on  testing

* Code review feedback, needs testing

* Added swagger

* Minor swagger fix

* Fix language

* Fixes:
- Added ClusterComponent.listActive which only returns non-deleted clusters
- Made cluster queries do a join to the label table, instead of doing a separate label select on _each_ returned cluster
- Add a LeoException to ProxyService
- Made CLusterDnsCache run on startup

* Fix comments

* Restore listActive() calls in ClusterComponentSpec

* Made ProxyService return appropriate status codes if the cluster is not found, or not ready

Added spec for ClusterDnsCache

* Update swagger with ProxyServer error codes

helgridly committed:
New-style PR checkboxes (#31)

ansingh7115 committed:
GAWB-2287: Cluster Initialization and support for GET (#28)

* GAWB-2287: support for GET

* review suggestions

* more review suggestions

* review suggestions

* more review changes

* more review suggestions

* fix tests

* moar review suggestions and refactoring

* adding site.conf

* remove unused imports

* fix tests

* some refactoring, lots of comments

* minor changes

* nitpick

* chanfe initializeBucket to for-comp and add a test

* fix test

* fix test really this time

* typo

* GAWB:2287 - most of the refactor

* some initial testing stuff

* adding test init actions file

* fix initializeBucket for tests

* +: doesn't seem to work here

* create bucket test

* mutability

* buckets are ArrayBuffers

* change mock bucketObjects to arrayBuffer

* remove redundancy in mock createCluster

* change contains to exists

* add firewall update back

* add testing for templating

* whoops import

* adding test for adding storage objects

* change for comp for initialize bucket

* change initialize storage objects to return nothing

* fixing tests

* annoying extra quotes

* jk

* make google projet unique

* change asserts to shoulds

* make mock return Future.successful(()), make ArrayBuffer into Set

* add more error handling

* add a more comments, bucket lifecycle

* Review suggestions

UrsaStutsman committed:
Us gawb 2354 (#34)

* GAWB-2354 adding git secrets hooks

* GAWB-2354 adding readmes

* Update README.md

rtitle committed:
[GAWB-2498] Notebook Service: poll Dataproc to get cluster IP (#33)

* Initial

* Try using EitherT

* most of the code written, needs testing

* Tests passing\!

* Start ClusterMonitor's on boot

* Fixes based on testing, working locally now

* Recover from Google 404 errors, and assume the cluster is deleted

* Minor tweaks

* Nicer way to construct a EitherT

* Fix swagger merge issue

* Push some logic into ClusterComponent to avoid duplication

* Minor comment updates

* another EitherT simplification

* Remove unused method

* Refactoring to move stuff into the Google DAO, WIP

* Cleanup

* Logging tweaks

* Fix incorrect catch statement in ProxyService

* PR feedback

* More OptionT comments

* Post-rebase fixes

* Latest PR feedback

rtitle committed:
Add log statement for failed Google requests for easier debugging

UrsaStutsman committed:
Update README.md

rtitle committed:
Use cookie-based auth for the Jupyter proxy (#35)

* Use cookie-based auth for the Jupyter proxy

* Fix status code in rejection handler

* Fix jupyter path in jupyter_notebook_config.py

* Better unauthorized error message

* Remove /api from jupyter path

* Pass in class name explicitly

rtitle committed:
[GAWB-2535] Notebook Service: support user-defined notebook extensions (#36)

* Wiring for extensionUri in the cluster PUT request

* Add install extension script to the dataproc config, and fix a concurrency bug

* Fix test

* Fixes based on testing

* Code review feedback

* Validate the jupyter extension uri in LeonardoService

* Fix pattern match

* Fix NPE

rtitle committed:
[GAWB-2002] Implement the GET /clusters endpoint to list notebook server clusters (#37)

* Working, just need additional tests

* Add tests

* Swagger

* Allow passing labels in a special way for swagger

* Swagger wording

* Comment

* More testing

* more tweaks

* Moved label parsing to LeonardoService, updated tests

* Handle _labels='' case

* Swagger indentation fix

* PR feedback, so far

* More swagger verbiage

helgridly committed:
GAWB-2613 Upgrade Hail version (#38)

vkumra-broad committed:
GAWB-2611 Add leo-demo app credentials (#39)

* Add firecloudServiceAccountCredentials

* Add leo-demo app credentials

* Fix variable name and type

rtitle committed:
[GAWB-2629] Delete the init bucket once the Leo cluster is ready (#41)

* Sanitize bucket name and delete bucket once cluster is ready. Tests not working yet.

* Fix tests

* Fix name sanitation logic

* Delete the init bucket if cluster creation fails

* Clear bucket before deleting it

* Comment fixes

* Code review comments, and use GCS parsing code from workbench-utils

* Change return type

* Add exclusions to workbench-libs dependencies

* Update workbench-libs to non-SNAP

ansingh7115 committed:
GAWB-2623: Endpoint to GET a list of clusters that are not deleted (#40)

* GAWB-2623: trying stuff

* oops:

* it works

* rebase

* adding direct db delete to test

* get list test

* changing to int

* fix direct db deletion

* it works - need to fix tests

* labels both ways

* make sure the db transaction completes before testing

* fix list clusters with labels test

* review suggestions

jmthibault79 committed:
Use value classes instead of strings
- other misc refactoring
- use newest workbench-libs
- Change ClusterInitValues to Strings
- rm ClusterResponse
- rm StorageObjectResponse

jmthibault79 committed:
Hello-World integration tests using the automation infrastructure
- needs major refactoring, see https://docs.google.com/document/d/1kAVXNoSYkt_BG6_Ej4-md-JWMlp414-dBGo8vX5uHOc/edit# for plan

helgridly committed:
add status enums to swagger (#43)

ansingh7115 committed:
GAWB-2624: adding build/push for leo notebooks docker image (#44)

* GAWB-2624: adding build/push for leo notebooks docker image

* oops

* double oops

* changing README wording

asingh7115 committed:
GAWB-2621, GAWB-2628: default cluster labels and dividing config

rtitle committed:
Potential fix for jupyter docker build (#48)

* Potential fix for jupyter docker build

* cd-ing into jupyter-docker in build.sh

* oops

* remove extra spaces

helgridly committed:
Save the init bucket in the db (#50)

helgridly committed:
fix for conf fail (#51)

rtitle committed:
Don't need to info log on every request (#52)

vkumra-broad committed:
GAWB - 2627 - Allow multiple clusters with same name (#49)

* Allow multiple clusters with same name

rtitle committed:
Update README.md

Added instructions for running unit tests.

davidangb committed:
update third-party dependency versions based on security scans (#53)

rtitle committed:
Add basic UserInfo checking to Leo. This is not actually used yet. (#54)

rtitle committed:
Set OAuth2 scopes on cluster creation, and don't send the sa key json to the cluster (#55)

jmthibault79 committed:
Swintegration update to work with QA and recent changes (#57)

- rm empty cluster list
- fail fast on creation error
- Use the standard leo SA because we don't need fancier scopes yet
- add DefaultLabels

jmthibault79 committed:
Run integration tests in Jenkins (#59)

ansingh7115 committed:
GAWB-2665: Whitelisting Endpoints (#56)

* whitelisting for cookie and noncookie auth

* clean up

* add whitelist to test config

* oops

* adding a mock userinfo

* it was not the way I thought it was

* oops

* some review suggestions

* add cache values to test conf

* typo

* add proxy test

* oops

* fix mock getEmailFromAccessToken

* handle illegal argument exception

* google token look up returns 401

* minor corrections

* more review suggestions - add token expiration handling

* deleting file added by mistake

* :whiteList to whitelist

* change executeGoogleRequestAsync back to what worked

rtitle committed:
Fix bug with cookie auth expiration (#61)

rtitle committed:
[GAWB-2648] Leo call sam to create pet service account at cluster creation time (#58)

* Added a SamDAO, mock+http versions, and unit tests

* Improvements to sam dao, and negative testing

* Basic sketch for LeoService calling SamDAO to get the pet sa

* Fix tests

* Added Sam config

* Rebase

* Add some logging to see wth is going on with tls

* Add dataproc worker role at cluster creation time, with some todos

* Fix comp error, more ugly hacks

* Tests need to compile too

* One of my ugly hacks had a typo

* Update libs version

* Needs a SNAP

* Update hash again

* Remove log statement

* Clean up todos

* Swagger: remove serviceAccount from ClusterRequest

* Remove debug statement

* Rebase and some, but not all, PR feedback

* Remaining PR feedback

* Remove unneeded test config

* Update libs dependency to non-SNAP

helgridly committed:
GAWB-2653 Set init bucket acl to user specified SA and leo only (#60)

jmthibault79 committed:
run integration tests in docker (#67)

rtitle committed:
[GAWB-2696] Leo: make /status endpoint (#65)

* First pass at Leo status endpoint

* Fixes

* Fix compilation error

* Update libs dependency

* Update swagger

* Update libs dependencies

rtitle committed:
Fix swagger indentation and improve StatusService logging (#69)

* Fix swagger and improve StatusService logging

* Debug log the OK case

ansingh7115 committed:
GAWB-2662: Configurable clusters (#63)

* GAWB-2662: adding ClusterMode and numberOfWorkers to cluster request

* config clusters - first attempt

* adding preemptible workers

* re-organize:

* fix test conf

* review suggestions

* minor

* change semigroup parameter names

* tests

* more test stuff

* fixing bug

* changing semigroup definition

* 1 worker cluster

* more semigroup changes

* oops

* clean up

* more cleanup sorry

* review suggestions

* review suggestions

* review suggestions

* clean up

* add minimum disk size test

* oops

* rebase weirdness

* review suggestions

* move clusterComponent test

* oops

* why can't I do anything right

* I want to go home

rtitle committed:
GAWB-2799, GAWB-2823: Leo multi project support (#68)

* Generate the SA token and propagate to the cluster

* Config fix

* Rename some things

* Just some commentary

* Update to latest libs

* Fixes to make things work across projects

* Temp fix, was passing the wrong project for the pet SA

* Config bugfix

* Rebase fixes; added ability to lookup a project number by name

* Fix test

* Grant bucket ACL to the COmpute engine default service account

* Add debug logging of the Google project

* Fix bug, wrong project was being used again

* Update wbGoogle to debug an issue

* Fix cluster deletion bug

* Comments

* Rebase fixes and PR comments

* Update model versions post rebase

* Update libs to non-SNAP

helgridly committed:
FCToken -> FCtoken (#70)

rtitle committed:
Fix NPEs in health checker (#72)

rtitle committed:
[GAWB-2818] Leonardo: Fix cookie auth expiration using a notebook extension (#71)

* Jupyter extension to refresh Google auth tokens

* Minor improvements

* Config oops

* cluster-docker-compose.yaml should not be a ctmpl

* Actually do the thing it's supposed to do

* js init values should be bucket paths

* Remove debug log

* PR feedback

rtitle committed:
Put pet creds on master and workers (#73)

jmthibault79 committed:
GAWB-2887 Set hadoop auth properties and Dataproc staging bucket ACLs (#74)

helgridly committed:
bad bucket! (#75)

jmthibault79 committed:
GAWB-2617 Automated Jupyter notebook tests (#66)

- Update WebBrowserSpec with download utils from UI Swatomation
- Changes required to make WebBrowserSpec run locally

davidangb committed:
upgrade jackson version (#76)

helgridly committed:
more than two executor instances, please (#77)

jmthibault79 committed:
Attempt to de-flake GAWB-2617 tests (#78)

* Increase timeouts
* download fix

rtitle committed:
Big libs update (#82)

jmthibault79 committed:
Use GoogleProject from workbench-libs (#84)

jmthibault79 committed:
GoogleServiceAccount -> WorkbenchEmail (#85)

rtitle committed:
Delete the SA key when the cluster errors out in ClusterMonitorActor. (#80)

* Delete the SA key when the cluster errors out in ClusterMonitorActor.
Add unit tests to cover GoogleIamDAO calls.

* Post rebase fix

* SWAT test for verifying pet credentials

* Post rebase fixes

* Remove Leo GoogleProject from automation

* Wording fix

* Don't hard code the number of keys

* Remove GoogleProject from ClusterKluge

* Attempt fixing swat

* Retest

* test fix, really

* Bump libs version to non SNAP

helgridly committed:
first pass leo blurb (#88)

vkumra-broad committed:
GAWB 2778 - Delete error clusters (#83)

helgridly committed:
GAWB-2895 Auth provider interface (#81)

helgridly committed:
link out to overview on wiki

jmthibault79 committed:
GAWB-2797: Make ready clusters immediately proxy-able (#87)

* GAWB-2797: Add ready clusters to DNS cache immediately

* withReadyCluster is no longer necessary

* Fix test compilation

* PR feedback: use Either instead of Future in ClusterDnsCache.processReadyCluster
Also some code cleanup in ClusterMonitorActor.handleReadyCluster

* Post rebase fixes

Commit:
Minor updates to repo documentation (#91)

* Add project status to README
* Add Verily to LICENSE
* Add placeholder for external contribution guidelines until we roll out our full new CLA process.

Commit:
Update LICENSE.txt

Remove a specific reference to Broad in favor of 'the copyright holder'

helgridly committed:
3 + 2 = fix

rtitle committed:
[GAWB-2942] Leo pets per project: Make ServiceAccountProvider interface and update Leo code to use it (#89)

* Initial refactoring. Main compiles, tests don't.

* Tests compile, don't pass

* Add DB migration, tests pass now

* Added Sam ServiceAccountProvider implementation, clean up some code

* Oops, fix name

* Beef up pet swat a little bit

* Rebase from develop

* Minor fixup

* post rebase fix

* Fix LeonardoModelCopy + ClusterKluge

* Add stub method to remove creds from instance metadata, if an override service account is used.

* Fix label swat test

* s/overrideServiceAccount/notebookServiceAccount/g

* Fix comment

* Fix substring check

* Dummy commit to test github

* PR feedback: made ServiceAccountProvider return both the Leo SA and the pem file

* Sleep a little more

* Jam a recover on removeServiceAccountKey so the delete doesn't fail

* Fix swat test flakiness, fix TODOs

* retrieving service account keys from Google is SUPER inconsistent and flakey

helgridly committed:
fix for minor swagger lies (#96)

rtitle committed:
Leo pets per project (#95)

* Implement new endpoint in HttpSamDAO. Fix some bugs as well

* Add PetsPerProjectServiceAccountProvider

* rebase oops

* Switch unit tests to PPP

* Update libs depdendency

* Stab at a swat test case, untested

* PR feedback, and actually switch unit tests to use a PPP implementation

* Fix swat label check and init script bug

* Fix spec bug; try NOT adding dataproc.worker to the pet SA

* Fix test, temporarily

* Revert "Fix test, temporarily"

This reverts commit 937abc0274596f520d3cd46adee40e638f7dd369.

* Put back dataproc worker

* Init script bugfix attempt. Committing to run SWAT

* More swat

* Need to set GOOGLE_APPLICATION_CREDENTIALS _conditionally_ in the docker container, not always

* Bash fail

* the env file needs to exist

* Should make swat pass

* Try and prevent spurious selenium timeouts

* Even more patience

* Increase selenium timeouts

* Update to non-SNAP

rtitle committed:
Multiline keys must be indented (#97)

rtitle committed:
Dataproc Worker concurrency bug (#99)

* Fix concurrency bug relating to removing Dataproc Worker IAM role from pets

* Spec

* Comment improvement

* Change method signature; add missing DB indexes

* Fix comment lie

helgridly committed:
added cluster creator to default set of labels (#98)

jmthibault79 committed:
Selenium upgrade to fix download issues (#100)

* also removed a false comment

helgridly committed:
GAWB-2386 stub /localize endpoint in Jupyter API (#101)

helgridly committed:
remove dependency on dsde-toolbox (#106)

helgridly committed:
GAWB-2386, GAWB-2407 loc/deloc (#104)

rtitle committed:
Ccross-domain cookie auth (#105)

* Cross domain cookie auth - make the Leo proxy accept a token via cookie OR Authorization header

* Make the cookie params configurable

* Add explicit login/logout endpoints per discussion with jmandel

* Add Google client ID param to the create cluster endpoint, per discussion with jmandel

* Change notebook extension to call gapi.authorize on a timer instead of init/signIn (thanks again jmandel)

* Post rebase fixes

* Change cookie logic to be more correct

* Minor fixes

* Remove googleClientId from the API and templating logic, in its own commit

* Stab at postMessage implementation to pass the client ID to the notebook extension. Swat still in progress.

* Temporarily comment out allowedOrigins to facilitate testing

* Revert "Temporarily comment out allowedOrigins to facilitate testing"

This reverts commit 73c1177b66cf07468da73b9299e9f986a965856d.

* PR feedback, and add swagger

* Need to set CORS headers for cross-domain cookies

* Fix unit test

* Support preflight OPTIONS requests for ajax

* ProxyRoutes bug

* Hopefully fix my proxy routes woes

* slight CORS fix, getting close

* Missed a couple cors headers

* Make postMessage data an object to be consistent

* Dummy page for testing. It works against a fiab :D. Needs swatification

* Oops, committed an .orig file

* JS tweaks and do cors for all proxy requests

* Fix tests

* dummy client fixes

* SWAT

* Debug error on Jenkins

* implement awaitLoaded in DummyClientPage

* Post rebase fixes

* Minor cleanup

* docker fail

* Check postMessage origin in dummy client

vkumra-broad committed:
GAWB 2608 - add AoU client library (#107)

* add AoU client library

vkumra-broad committed:
GAWB 3026 - give clusterserviceaccount access to dataproc bucket (#111)

* give clusterserviceaccount access to dataproc bucket

rtitle committed:
s/FCtoken/LeoToken (#109)

* Change cookie name: FCtoken -> LeoToken

* Just check the origin

jmthibault79 committed:
GAWB-2838: BigQuery inside notebooks (#113)

* Add Bigquery scopes to Dataproc cluster SA
* BigQuery Swat test
* Update formatting for new version of executeCell()
* Copy Google Role endpoints from UI-Repo Swat

jmthibault79 committed:
GAWB-3023 Create new billing project for every new-cluster test (#114)

* Bump handleResponse/awaitResponse times to 1 sec and billing account creation to 10 min

rtitle committed:
Slimmer Jupyter docker image (#119)

* Dockerfile reduction and refactoring

* docker seventeen

* Split Dockerfile into separate layers for prereqs + Java + Spark/Hadoop/Hail + pips

* Dockerfile updates based on feedback

* I think I need to mount Hive

* More spark confs that seem to help

* More tries

* useless Java option

* Reduce spark.yarn.am.memory

* Link JIRA in comment

rtitle committed:
GAWB-3082: Fix intermittent unit test failure in ProxyRoutesSpec (#124)

rtitle committed:
Fix ProxyRoutesSpec for real (#125)

* Add println

* Increase route timeout

ansingh7115 committed:
GAWB-2896: Do Sam Auth (#117)

* initial - probs doesn't work

* make the notebook action strings correct

* fix execution context bug

* try a different way of authing

* i dunno

* logging stuff

* fiab stuff

* trying to figure out why resourceAction is giving me a false

* hmmm - are we actually getting a boolean and passing in the right stuff?

* remove whitelist D:

* approximately

* import statements

* stuff

* add back recreate cluster

* fixing samauth test

* add cache stuff to config

* make it actually talk to Sam

* oops

* use a real fake pem

* ignore test for now...

* forgot this

* update for sam changes

* resolve conflicts

* let's not be too clever

* tiny correction

* log some stuff

* will this work idk

* decode

* logging stuff

* more logging

* add back whitelist + update for sam changes

* do it right

* log more things

* maybe, idk

* omg this might work i think

* no, but maybe this?

* wild guesswork

* this has to be it

* as json this time

* try this

* idk

* oh this actually worked

* comments and test stuff

* switch auth for tests

* src

* im sleepy

* some initial review changes

* Change strings to model classes

* change tests

* oops broke tests

* new auth provider interface

* fix whitelist

* fix reference.conf

* fix sam server url

* use mockLeoAuthProvider in LeonardoServiceSpec for now

* go back to whitelist

* sam auth provider spec

* list clusters courtesy of Rob and Hussein

* override canSeeAllClustersInProject

* hussein's nitpick + executionContext

* bracket

* samclient prelim

* move project and notebook actions back to SamAuthProvider

* [nomerge] Better LeoService interop w/ auth providers (#120)

* new auth provider interface

* wip

* better

* commentary

* tests for optimized leo list clusters

* whitelist fixes

* mockito test

* removing extra stuff from commontestdata

* silly

* fix tests

* fix tests for real

* maybe don't use mockito

* add billing project manually to mock

* tiny mistake

* maybe this

* I can't seem to do anything right

* now we're getting somewhere

* wait no, add back mockito

* why can't this be over

* what's in a name

* stop reading my commit messages, the pressure to entertain is too much

* I miss the days when my commit messages were just for me

* please pass

* clean it up make it pretty

* get and contain issue

* that which we call a rose
By any other word would smell as sweet but would probs be confusing

* change artifactory

* review suggestions

* why would I get it right the first time

* something broke again

* fix for notebookclusters too

* stupid race conditions

* change cacheExpiryTime to FiniteDuration

* fix reference.conf cacheExpiryTime

* ergh

* missed one

* maybe this

* I think I'm getting closer

* well, maybe

* update sam swagger client version

* so muh simpler

* oops naming resources

* remove logging

* fix the automated tests

* have more patience

* typo

* remove logging messages

* do the simple thing first

* fix the owner delete bug

* don't need userEmail in internalDeleteCluster anymore

* pass in full cluster to notify methods

* fix tests

* go back to using email project and cluster name

* refactor swat tests a bit

* missed two...

* change it again sigh

* fix tests

helgridly committed:
GAWB-3041 make notebooky tests reuse the same cluster (#121)

* Make notebooky tests reuse the same cluster
- new LeonardoTestUtils trait
- Run ClusterMonitorSpec in parallel
- Don't need to explicitly mixin WebBrowserSpec
- centralize hail file vals
- Ensure withOpenNotebook/withNewNotebook always terminate their kernels
- Removed redudant test which uploads but doesn't execute notebook

UrsaStutsman committed:
Moving automation into workbench libs (#134)

* use workbench libs

* auth token for leo, other deletions, tweaks

* adding methods to configs

* adding some more info

* forgot to render users.json

* build/push swat tests docker image in jenkins build script

* all up to date

* syncing users json

* Add campaignManagers to users

* workbench-libs needs these (TODO remove)

* fix for app conf

* real fix for app conf

* Hermione and Ron

* more app conf fix

* more app conf ctmpl

* Update scalatest version for workbench-libs compatibility

* implicits are the devil

* dammit weasley

* MORE PATIENCE

* fix build script when pushing to develop

* save git SHA during build script

rtitle committed:
Update README.md

Added link to the production swagger UI to README.md

helgridly committed:
configurable providers are live now

rtitle committed:
Leo/Sam auth: handle Sam API exceptions in SamAuthProvider (#138)

* Add error handling to providers

* Add tests

* Clean up

* Add logging

* Exception recovery not needed on SamAuthProvider

* Better specs and error handling

* Add comment

* Log text nit

* Add comment to specs

jmthibault79 committed:
GAWB-3108 Defer Deletion of Sam Cluster Resources (#139)

* Delete cluster resources in Sam only after deletion is complete in Google and the Leo DB
* Integration Tests: Downgrade selenium/node-chrome to 3.8.1-erbium

jacmrob committed:
add docker tag for tests repo (#174)

jacmrob committed:
I screwed up and changed the wrong docker tag

jmthibault79 committed:
withNewGoogleBucket and withNewCluster / createNewCluster updates (#176)

* config shenanigans

ahaessly committed:
Ah leonardo 102: comment config files (#171)

* update configs; add example
* break out certificate docs into separate md file
* move example file to top level dir

ahaessly committed:
update test dependency (#177)

set version of serviceTest workbench libs to fix for register cleanup

vkumra-broad committed:
GAWB 2839/Issue 140 - Create new dataproc staging bucket per cluster (#169)

* create new staging bucket per cluster

eap committed:
docker/build.sh; update flags, allow GCR images. (#172)

* docker/build.sh; update flags, allow GCR images.

Make it easier for users of Google Container Registry to build
images. Also make the command line parsing a little less sensitive
to order and less dependent on environmental vars for config.

* Add missing "$" to PRINT_HELP bool.

* Fix use of PROJECT and BRANCH vars

helgridly committed:
add missing log to leo docker build (#180)

ansingh7115 committed:
LEO-126: Allow users to provide a script that runs in their jupyter docker (#170)

* initial commit - it compiles, so that must mean it works - add a jupyter user script to everything

* add a column to the db for the user script uri

* refactor better

* tiny

* oh my god google uris cant have capital letters why did u forget this

* be more careful

* make the init actions print some stuff

* change logging

* just trying stuff

* still trying stuff

* maybe sudo

* try -it bash

* try chmod

* try not doing it in the background

* promising - maybe sudo

* the docker container doesn't have sudo, trying execing as root instead

* do it right

* try more stuff

* i think it was me being stupid

* actually, I did need the chmod

* clean up

* make error message more generic

* try running user script before pyspark

* try -d

* it works yay

* specify gcs uri and fix yaml

* fix tests - use CommonTestData for leo service spec

* rebase confusion

* prelim automated tests

* user script test - hopefully bucket permissions are alrigh

* use the already there swat test bucket

* fix object type

* oops

* didn't need object type

* maybe empty string instead of null will work

* okay, can't use null for some reason, explicity tell it we're using a shell script

* try abstract input stream + plain text

* clean it up, test calling the library, not just importing it

* forgot to add userScript in places...

* maybe the bucket name isn't being sent right?

* refactoring cluster and label checks

* add the labels

* oops

* don't test the import step

* make a new cluster, and test without the user script

* lowercase

* some clean up

* review suggestion - fix swagger description

jmthibault79 committed:
Resurrect and update PetServiceAccountProvider (#181)

- Use keys in the cluster's project, not leo's
- map JIRAs -> issues
- Update label check to be ServiceAccountProvider-agnostic

rtitle committed:
Add back Jupydocker ENTRYPOINT 2 (#187)

* Add back pyspark ENTRYPOINT

* Try bash an an entrypoint

* temp don't delete failed cluster

* new entrypoint

* A better entry point

* Revert "temp don't delete failed cluster"

This reverts commit b387a64a1d46bf808d438e1ff8b262317561f41e.

* A different way of doing it

* Better comment

rtitle committed:
Use StorageDAO in Leo, plus other refactoring (#182)

* Split storage functions out of DataprocDAO (#146)

* Get things working again, and break out a BucketService class

* Self review and cleanup

* Fix config name

* It was still setting ACLs in parallel

* Forgot to update serviceTest dependency

* Was setting the wrong bucket in the Cluster object

* Nuke GoogleExceptionSupport; throw WorkbenchExceptions from HttpGoogleDataprocDAO and handle them elsewhere

* PR feedback: rename BucketService to BucketHelper

* Rebase with anu

* Rebase with joel

* Remove foo

jacmrob committed:
set up junit test reports (#179)

jmthibault79 committed:
Change getPetOrDefault() to return all available SAs (#192)

jmthibault79 committed:
Preemptible test for #127 (#178)

* update pet credential test for PetNotebookServiceAccountProvider

rtitle committed:
Make SamAuthProvider more performant (#190)

* Add cache of notebook auth results and enforce Sam timeouts

* Put sam providers in sam sub-package

Add specs

* Fix wording

* expireAfterWrite, not access

* maybe make it 30 seconds

* Minor comment fixes

* Cache canSeeAllClustersInProject as well since it's called by list clusters

* Minor PR feedback

* Use CommonTestData

ahaessly committed:
Ah debian upgrade (#196)

* upgrade to stretch
* missing gnupg, procps

jacmrob committed:
refactor local run scripts to use common automation repo; add documen… (#201)

rtitle committed:
Ignore flakey error test (#210)

* Ignore flakey error test

* ignore the other flake too

jacmrob committed:
set leonardo-tests latest image to env (#212)

ansingh7115 committed:
LEO-129: Add a Python 3 Kernel (#195)

* see if it just works?

* do python 2 explicitly

* try 2.7

* add python 3 commands

* make it non interactive

* add pip3 packages

* forgot a backslash

* specify python3-dev

* take out python2-dev for now

* unsure why we need ENV USER, move it somewhere else and see if anything bad happens

* actually, it's just a warning

* use apt-get to install

* seaborn - use pip

* back slashes

* cant pip install libz-dev

* add python3-dev

* try installing zlib1g

* try re installing libz-dev

* install python-lzo

* liblzo2-dev, who knew rite

* add automated test

* pip3 installing the decorator does weird things...

* forgot you have to specify pip2 or pip3 in user script test

* some review suggestions - still have to fix up spark

* prelim script for testing on fiab

* forgot entrypoint

* doing apt-get clean too early

* this works I think

* I think this really should work

* stupid effing backslashes

* this works, its different this time I swear

* didn't refactor correctly

* rebasing for debian gave me two different python3s

* actually we needed that line - i think it sets the worker python version

* spark python 3.4 vs 3

* debian stretch doesn't have 3.4, maybe workers should be 3.5

* make worker nodes install python 3

* try using pyenv

* specify pyenv'

* I think we do want to export the path

* does changing the header thing do anything

* switched the wrong one

* this works

* livin fast and dangerous

* go back to jessie this is too weird

* change back sources list

* add a spark test to automation

* clean up

* review suggestion

* more clean up

jmthibault79 committed:
Specify which repo we're checking out master in (#213)

ahaessly committed:
leo132 - add R kernel (#211)

* add R kernel
* chown of .local
* add sparkr
* add tests

vkumra-broad committed:
Issue - 141- Save error details for errored clusters (#186)

Save error details for errored clusters

jmthibault79 committed:
Make git secrets hook executable (#224)

vkumra-broad committed:
[205]Intermittent test failure: failed to delete error'd cluster (#219)

Updated the error message check when deleting a cluster. Re-enabled failing test.

jmthibault79 committed:
Re-enable test for LEO-204 (#223)

jmthibault79 committed:
Update to latest service test: has gpalloc and thurloe fixes (#226)

jmthibault79 committed:
Ignore test.  See issues #204 and #228 (#229)

rtitle committed:
#222: Fix Hail (#231)

* #222 Hail integration is missing dependencies
#221 jupyter_install_notebook_extension.sh script is misnamed in the init script

* Hail tutorial test caes

* Fix upload download tests

* Remove foos

* Fix failing test

* Add quick check for clsuter status

jmthibault79 committed:
Use GPAlloc (#235)

ansingh7115 committed:
clustermonitertimeout (#240)

rtitle committed:
Peg tornado at version 4.5.3 (#241)

* Peg tornado at version 4.5.3

* Peg python3 kernel as well

eap committed:
GCE configure and deploy script (#203)

* Add gce deploy script and config

This change adds deploy.py which is mostly
successful at configuring Leo on a bare GCP
project. There's still some bugs to be worked
out; primarily an auth issue block access to
the notebook proxy.

* Some trivial code cleanup for readability.

* Configs are now owned by ubuntu user.

* Script user fix and service acct passwords.

1) The init script used 'su' rather than sudo bash << EOF to switch
   user. This caused some commands to not be executed.
2) Service accounts have a limit on the number of user-managed keys,
   I've added a check for existing keys that allows users to delte
   keys during deploy (thus preventing a breaking error during the
   last stage of deploy).

* Remove * from cluster domain template.

* Change common name for leo client.

* Added working site.conf

* Remove unused volumes in docker compose.

* Remove warning, ready for review

* Rebase PR, and allow GCR notebook image

The notebook image had been taken from the prod
leonardo docker repository. The recent changes
to notebook kernels had caused breaking changes
in the startup script. By pinning the notebook
image to the built images from this deploy
script, we now ensure that the launch script
is compatible with the target images.

* Code review fixes

rtitle committed:
Wait on futures in preemptibles test (#238)

* Await GCS futures

* Patience and un-ignore

* Increase hail timeout

rtitle committed:
Retry apt-get update in init script because it's flakey (#244)

rtitle committed:
Clean up mysql containers (#246)

rtitle committed:
Auto tests: log cluster errors and save init script logs as build artifacts (#245)

* Download log file from the staging bucket in tests

* Clean up

* change log dir

* Specify a log suffix

* Add a clue

* Change log dir

* Move some stuff

* Post rebase fixes

* Change output file name, it might be hitting Jenkins exclusions

* Change download file again

* More filename fixes

* It might work if it's in target since that is already mounted

* This time it's gonna work

* Might as well grab both init and startup logs

rtitle committed:
Stop leaking billing projects in Leo tests (#248)

rtitle committed:
Init script timestamps (#249)

* added timestamp logging

* Make a log function

* Typo

rtitle committed:
Remove AoU client library because it's outdated (#250)

rtitle committed:
We do not import the AoU library by default anymore (#251)

jmthibault79 committed:
Exception type changed (#267)

rtitle committed:
Leo creating too many pets (#270)

* Log sam APIs

* Leo creating too many pets

* Scaladoc

* Debug fixes

* PR feedback

rtitle committed:
Improve cluster cleanup-on-error logic (#273)

* Improve cleanup-on-failure which was causing clusters to be stuck in Creating status

* Formatting fix

rtitle committed:
Users unable to install R packages (#278)

* add R packages

* add backslash

* update

* testing

* testing

* Remove comments

* Need to chown the .rpackages directory

* Add automated test case

rtitle committed:
Issue 272: CORS headers are double-set for proxied Jupyter rest API (#276)

calbach committed:
Include git in Leo docker image (#266)

dvoet committed:
upgrade service test libs (#269)

* upgrade service test libs

* Grab the latest serviceTest, and make google a non-SNAP

ansingh7115 committed:
LEO-214: Fix Downloaded Notebook Name (#264)

* update akkahttp

* update akka

* filter headers

* it works

* clean up imports

* review suggestions

* adding tes

* fix automated test

calbach committed:
Implement a delocalizing contents manager for GCS (#259)

* Fix run-jupyter-local to use the Leo docker image

* Implement a delocalizing contents manager for GCS

* Add tests

* rename and cleanup

rtitle committed:
Really unignore preemptibles test (#252)

vkumra-broad committed:
Issue 220 : Leo SA should not need permissions on user scripts (#253)

* Update Leo Service to validate user script using the cluster SA

* added some logging

* more logging

* change to useremail

* create bucket to see what it's acls are

* create better bucket

* more logs

* what happens if I change it back to leo Pem

* Leo pem creds work

* add scopes

* update to read only scope

* fix test and clean up

* clean up

* remove debug code

* unit test fixes

* Improvements for obtaining a token-based DAO (#256)

* Improve test and add recover

* uncomment code

* Post rebase fix

ahaessly committed:
GAWB-3216 add logging for notebook calls (#261)

* update docs

* more doc

* fix db name in build.sh; parameterize

* remove unnecessary comment

* add missing quote

* update repo url

* add log command

* add more logging

* add log for metrics

* add logging for notebook requests

* add logging for notebook requests

* Revert "add missing quote"

This reverts commit 488efcd00b8e85f76cb95d08c8594f5888ec7102.

* Revert "fix db name in build.sh; parameterize"

This reverts commit 36bfa0e31a6371209d33952222569d6de0efb9f8.

helgridly committed:
Put the init/staging bucket label at the beginning of the bucket name (#279)

dvoet committed:
Pet name change (#286)

jacmrob committed:
tag test image with git hash when building real env (#287)

vkumra-broad committed:
Add test for jupyter extensions (#285)

jmthibault79 committed:
Increase R package install test timeout (#292)

* Update executeCell() to verify the rendering of the last cell

ansingh7115 committed:
Make the websocket flicker go away (#297)

vkumra-broad committed:
Reduce time for cluster create request (#294)

rtitle committed:
290: Add retries to SwaggerSamClient (#299)

* 290: Add retries to SwaggerSamClient

* Wasn't testing the right thing

calbach committed:
Improve delocalize docs (#301)

rtitle committed:
Fix selenium kernel shutdown condition (#304)

rtitle committed:
Increase LeoRoutesSpec timeout (#302)

jmthibault79 committed:
More verbose errors (#305)

ahaessly committed:
change leo db name (#233)

* change leo db name

* fix db name in build.sh; parameterize

* add missing quote

* Hard code CONTAINER2

rtitle committed:
Pause/resume clusters (#260)

* Pause and resume clusters

* Improve handling of concurrent stop/starts; don't need to add/remove IAM roles for stop/start

* Test fix due to changes in behavior of executeCell

* Increase start patience; thought I did this before

kyuksel committed:
Automatically set cookie for all /api endpoints (#310)

kyuksel committed:
Fix for cookie age unit bug (#313)

rtitle committed:
Synchronous jupyter_localize_extension (#309)

* jupyter_localize_extension.py: add synchronous mode and better input sanitization

* tornado better

* Update swagger and beef up localization automation tests

* Improve loc/deloc tests; they should pass now

* PR feedback, and made sync mode the default

* variable name change

rtitle committed:
Upgrade Java version and add --ignore-installed (#325)

* Upgrade Java version and

* downgrading decorator seems to fix it

rtitle committed:
Allow data URIs in jupyter_localize_extension. (#307) (#323)

jmthibault79 committed:
Update service-tests to 0.8-cd592ce (#318)

eap committed:
First UI pull request (#314)

* User Interface: template React app and error handler.

* User Interface: added google auth

* User Interface: cluster state and cards.

* User Interface: update package files

* User Interface: app-bar icon and status icon

* User Interface: Add connect button

rtitle committed:
Set staging bucket ACLs in a separate call (#335)

eap committed:
Owner email added to jupyter environment (#333)

eap committed:
jupyter-docker: move to R 3.4 (#334)

kyuksel committed:
Capability to set project team level ACLs

rtitle committed:
Pip install as user by default (#345)

* Pip install as user by default

* Cell numbers start at 1 after the kernel is restarted

* This seems to be a bit cleaner

* Fix test failures

* Minor test cleanup

rtitle committed:
Do the pip install verification in a new notebook (#347)

* Do the pip install verification in a new notebook

* Use the same NotebookListPage instance

* Revert "Use the same NotebookListPage instance"

This reverts commit 5743a95dea03dedf996cde0afe96472c7de4e862.

* Save notebook to avoid popups

helgridly committed:
GCR for leo-notebooks image (#291)

* switch to gcr

* Parameterize notebook repo

* missed one

* Fix typo in reference.conf

eap committed:
User interface: dynamic status icon. (#327)

* User interface: dynamic status icon.

Status icon for better "quick glance" cluster status
recognition. See it in gif form here
https://docs.google.com/document/d/1_EDQR347f7CsTxy5OWiXOn3iV9IKwOv-M9wPmilWUlU

* Added new starting/stopping/stopped status

eap committed:
User Interface: add delete button / dialog. (#328)

This change adds a delete button to each cluster card. that
button will create a confirmation dialog when clicked, it will
manually set the status to "Deleting" when the delete operation
returns with success.

See the new component in action here:
https://docs.google.com/document/d/1dcN4Edbuiny8e9CIEp4KrY5AZGIT9ULWAuHM78BrdkA

eap committed:
User Interface: create cluster modal and form (#329)

* User Interface: added create cluster form

* Fix field description color

* Add more machine types

rtitle committed:
Better cluster status checking (#331)

* Make SamAuthProvider.notifyClusterDeleted recover from 404s

* initial

* Add auto tests

* Some test restructuring

* Fix status transitions

* Fix stop/start exceptions

* Fix unit test

* Need to make sure all deletions are monitored

* Update the swagger

* Fix variable name

* Add an exploratory Thread.sleep to debug a test failure

* TEMP add logging to debug test failures

* Compiling is hard

* Remove debug logging

* Maybe the nested FreeSpec was confusing things

* TEMP add debug logging. Will revert

* Try adding some patience for get-after-create

* Remove debug logging and fix deletion monitor

* PR feedback

* Post rebase fix

* PR suggestion

* Remove ParallelTestExecution and add debug logging to tests

* Add back ParallelTestExecution

* Improve SamAuthProvider logging and increase retries

* Try creating a billing project in beforeAll

* Revert "Try creating a billing project in beforeAll"

This reverts commit 817a91a865d22ef7dd7b38aeae7cc698e8addde2.

* Invalidate the cached pet token when claiming a new gpalloc project. It's ugly but should resolve the test issues.

* Remove phanton endpoint, disable the token cache in fiabs instead

* Use actual ficus method

* Don't call endpoint that doesn't exist

* PR feedback

rtitle committed:
[#311] Add a "create in stopped state" endpoint (#349)

* Better cluster status checking

* Better status checking

* Initial impl of stop-after-create

* Fix merge issue

* Clean up and add an automation test

* Typo, ceated

* Post rebase fixes

* Fix issue noticed in review: Error'd clusters should not be stoppable

* Fix test compiliation error

rtitle committed:
[350] Cannot localize notebooks with spaces in the file name (#351)

* Basic test case

* Remove pipes.quote

* Fix async boolean parameter parsing

* Test fixes

* More test fixes, should pass now

calbach committed:
mv gsutil delocalize to the background (#353)

kyuksel committed:
Separate Python and Spark kernel installations (#352)

rtitle committed:
Make stopAfterCreation truly optional in the create cluster request (#356)

* Make stopAfterCreation truly optional in the create cluster request

* Make it default to false in swagger

* Fix unit test

kyuksel committed:
User script to hide spark kernels (#357)

aweng98 committed:
sbt forking jvm (#355)

* Update Dependencies.scala and build.sbt

vkumra-broad committed:
Issue 130: Support multiple jupyter extension in Leo (#346)

rtitle committed:
Peg the statsmodels dependency for compatibility (#371)

eap committed:
User Interface: new options and other fixes (#370)

* Card index now causes refresh on state change.
 * List cluster request filters properly.
 * New options.

rtitle committed:
Add support for obtaining a GCR service account from vault (#363)

* Add support for obtaining a GCR service account from vault

* Remove debug line

* Add default SERVICE_ACCOUNT_KEY_FILE

rtitle committed:
update README

rtitle committed:
Revert "update README"

This reverts commit c57f29c2c8912127111bc363ff6413df023777a9.

calbach committed:
Fix Swagger validation errors (#373)

eap committed:
Explicitly set VPC net/subnet on cluster creation (#365)

aweng98 committed:
update lib versions (#376)

* update lib versions

* workbench-util ver

vkumra-broad committed:
Issue 367: Auto Freeze - Phase 1 (#374)

* Initial commit

* fix swagger

* PR feedback

* PR Feedback 2

* fix swagger

rtitle committed:
Fix VPC configs (#377)

* VPC config refactoring

* Fix comment

* ADd newline

vkumra-broad committed:
Issue 368 - Auto freeze phase 2 (#381)

vkumra-broad committed:
Issue 369: Auto Freeze Phase 3 (#390)

vkumra-broad committed:
Date accessed should only be updated when cluster is ready (#396)

rtitle committed:
Rt init scripts in foreground (#393)

* Install init scripts in the foreground

* Do even less in the background

* Don't install extensions as root

* pip install as root, jupyter commands as jupyter-user

* Install sudo

* Log jupyter server output to a file

* Fix typo

* Fix redirect

* Forgot to do 1 of them as root

* Actually don't install it as root

* Use nifty bash trick

* Preserve environment variables in sudo

* --user

* Add mlr test, it works

* Revert "Add mlr test, it works"

This reverts commit 8e33738b808c1493d25dac76485bc08841201774.

rtitle committed:
[Issue 398] Install libxml2-dev which is a system requirement for some R packages (#400)

* Install libxml2-dev which is a system requirement for some R packages

* apt-get update as well

* Add mlr test, it works

* Oops, compile error

* Fix mlr install check

* Piggback: set LC_ALL ev for pretty histograms

* Missing quotes in test assertion

jacmrob committed:
fix render-local-env script for using local UI (#403)

rtitle committed:
Fix locale so unicode characters actually display (#405)

* Fix locale so unicode characters actually display

* Fix assertion due to locale change

rtitle committed:
Add Google Cloud Source Repositories scopes to the pet SA on the cluster (#409)

* Add Google Cloud Source Repositories scopes to the pet SA on the cluster

* Set READ_ONLY scopes per PM

akarukappadath committed:
Min disk (#411)

* changed min Disk size to 10GB

* updated default min disk size to 100

* added migration to changelog

* added units to description of disk size

aweng98 committed:
Fix sbt eviction warnings (#412)

* first pass

* akka

rtitle committed:
MLR takes a while to install (#421)

aweng98 committed:
fix eviction warnings (#419)

mistake

jmthibault79 committed:
rm the copypasted page hierarchy (#427)

jmthibault79 committed:
Update Scala version to 2.12 (#428)

rtitle committed:
Add example notebooks to git (#433)

* Add some example notebooks to github with a README

* Update README.md

* Update README.md

* Add upload image

* Update README.md

* Update README.md

* Update README.md

rtitle committed:
Update README.md

jmthibault79 committed:
Enable independent config values (#430)

vkumra-broad committed:
Issue 385: Clusters not immediately proxyable on restart (#425)

akarukappadath committed:
setCookie throws 500 error instead of 401 when token is invalid/expired (#423)

* changed 500 error to 401

* debugging

* throws 401 error when token is illegal

* adding automation tests

* fixed automation test

* cleaning up push

* updated Config to LeonardoConfig

* removed users.json from gitignore

* using Students instead of NotebooksWhiteListed for testing

* adding users.json back into gitignore

* removing users.json

* changed credentials to work with automated testing users.json

akarukappadath committed:
catch 403 before it retries

aweng98 committed:
Break up NotebookInteractionSpec (#429)

* NotebookPyKernelSpec parallel

1st try

fix name

additional fixes

more fixes

fix

fix

fix

parallel NotebookRKernelSpec

NotebookInstallSpec

addressing feedback

logger get details

* comment out delete log

akarukappadath committed:
changed models to accomodate autopause specification

akarukappadath committed:
update model

akarukappadath committed:
Leo retries 403 error until it times out (500) if Dataproc is disabled (#436)

* catch 403 before it retries

* using new exception DataprocDisabledException

* catch 403 before it retries

using new exception DataprocDisabledException

* fixed error from merge

rtitle committed:
Experimental: allow opening Jupyter in an iframe (#442)

* Experimental: allow opening Jupyter in an iframe

* Per discussion, remove single-tab mode. Keep Content-Security-Policy.

aweng98 committed:
should include (#447)

rtitle committed:
Fix JupyterLab and related refactoring (#441)

* Refactor jupyter scripts and fix lab installation

* Fix chmod command

* Fix kernelspec args

* Fix more setup issues

* Add newline

* Add basic JupyterLab test

vkumra-broad committed:
Use commontestdata and refactor testLeoRoutes (#448)

eap committed:
User Interface: deployment using dockerized static site (#443)

* Update UI deployment.

* Add a pre-loading config system for the UI that works
  when being served from a static server (like nginx).
* Add a build system for the UI that creates an env
  controlled container.
* Allow additional scopes to be added during auth.
* Update deploy script to show how the new front-end
  container is used.

* Fix newlines

* Convert config-loader to a react component

* update config comments.

* Review response: method definition in ConfigLoader.js

eap committed:
User Interface: cluster cards manage state refresh on their own. (#452)

Prior to this change, cluster cards would need to be refreshed manually with the refresh button in the UI. This change ensures that clusters whose state is in flux (deleting, creating, starting, stopping, etc...) will hit the Leonardo API periodically to check for a new state.

This change also adds the behavior that running clusters will re-check their state every 5 minutes to make sure that their state hasn't changed (for example, due to auto-suspend).

eap committed:
Add IamProxyAuthProvider. (#450)

This new provider uses client credentials to test permissions
on a project before granting access to a user. Thus, it is
proxying the user's IAM permissions.

eap committed:
Build script fixes. (#454)

* Add a "-ui" option in the main builder script to
  allow optional building of the UI docker image.
* Reduce verbosity of wget in the jupyter-docker build.

eap committed:
Update react-material, add config script input field. (#455)

* Add the field jupyterUserScriptUri to the create clusters from.
* Updated material UI to from 1.0.0.beta43 to 1.2.3.

akarukappadath committed:
debugging

eap committed:
UI: fix bug: null jupyterUserScriptUri fails to create cluster. (#456)

Add logic to determine if the value is set before applying it to the cluster model.

akarukappadath committed:
bug in ClusterComponentSpec

akarukappadath committed:
tests all working

akarukappadath committed:
updated swagger

akarukappadath committed:
fixed some conflicts

rtitle commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -618,6 +620,13 @@ definitions:
         description: |
           The date and time the cluster was last accessed, in ISO-8601 format.
           Date accessed is defined as the last time the cluster was created, modified, or accessed via the proxy.
+      autopause:
```
These are in `Cluster` but I think they should be in `ClusterRequest` instead? Currently in Leo `Cluster` is the API _response_ for put/get/list cluster requests.

rtitle replied:
Looks like `ClusterRequest` should have `autopause` and `autopauseThreshold` and `Cluster` should just have `autopauseThreshold` right?

rtitle commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -618,6 +620,13 @@ definitions:
+        description: Whether autopause feature is enabled for this specific cluster.
+      autopauseThreshold:
+        type: int
+        description: The number of minutes to elapse before the cluster is autopaused. If autopause is set to false, this value is disregarded. A value of 0 is equivalent to autopause being turned off.
```
Good text - only comment is to make it clear that the cluster only auto-pauses after it's _idle_ for the threshold. Maybe "The number of minutes of idle time to elapse..." ?

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:
```diff
@@ -75,6 +80,7 @@ trait ClusterComponent extends LeoComponent {
     def serviceAccountKeyId =         column[Option[String]]    ("serviceAccountKeyId",   O.Length(254))
     def stagingBucket =               column[Option[String]]    ("stagingBucket",         O.Length(254))
     def dateAccessed =                column[Timestamp]         ("dateAccessed",          O.SqlType("TIMESTAMP(6)"))
+    def autopauseThreshold =                  column[Int]             ("autopauseThreshold")
```
Super nitpick but it's nice to try and align the fields :)

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:
```diff
@@ -252,9 +258,25 @@ trait ClusterComponent extends LeoComponent {
+//        .filter(record => tsdiff(minute, now, record._1.dateAccessed) > record._1.autopauseThreshold)
+//        .filter(_._1.status inSetBind ClusterStatus.stoppableStatuses.map(_.toString)).result map { recs => unmarshalClustersWithInstancesAndLabels(recs)}
+//    }
+    def getClustersReadyToAutoFreeze(): DBIO[Seq[Cluster]] = {
```
Cool - this works? I'm glad :)

Remove the commented code above.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:
```diff
@@ -263,7 +285,7 @@ trait ClusterComponent extends LeoComponent {
       } else {
         // The trick is to find all clusters that have _at least_ all the labels in labelMap.
-        // In other words, for a given cluster, the labels provided in the query string must be
+        // In other words, for a given cluster, the labels provided in the query strincg must be
```
typo

rtitle commented on src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20180625_autopause_threshold.xml:
```diff
@@ -0,0 +1,11 @@
+    <changeSet logicalFilePath="leonardo" author="azmina" id="autopause_threshold">
+        <addColumn tableName="CLUSTER">
+            <column name="autopauseThreshold" type="INT">
+                <constraints nullable="false" />
```
Hm, since this is a new not-null column we should probably include a migration to update all existing clusters in the DB. Something like:

1. Add the new column as nullable
2. Update all existing clusters to the current default
3. Make the new column not-null

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -53,8 +53,8 @@ case class InitializationFileException(googleProject: GoogleProject, clusterName
 
-case class BucketObjectAccessException(userEmail: WorkbenchEmail, gcsUri: GcsPath)
-  extends LeoException(s"${userEmail.value} does not have access to ${gcsUri.toUri}", StatusCodes.Forbidden)
+case class BucketObjectAccessException(userEmail: Option[WorkbenchEmail], gcsUri: GcsPath, errorMessage: String)
```
Is this relevant to this PR? I feel like this was maybe carried over from a different change? Should probably revert it.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -358,9 +359,14 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
+      autopauseThreshold = clusterRequest.autopause match {
+        case None => autoFreezeConfig.autoFreezeAfter
+        case Some(false) => 0
+        case _ => if (clusterRequest.autopauseThreshold == None) autoFreezeConfig.autoFreezeAfter else clusterRequest.autopauseThreshold.get
```
Can we sanitize against other potential bad inputs? What if we get:
```
autopause = Some(true)
autopauseThreshold = Some(-2452345)
```
Currently we'd pass this directly to the DB query and bad things might happen. `Math.max(0, autopauseThreshold)` would probably resolve this particular issue.

Otherwise, the logic looks good, I think.

rtitle reviewed (changes requested):
See comments above. In general this looks great though - nice work!

rtitle committed:
Ignore flaky preemptible test (#464)

kyuksel committed:
Async Cluster Creation - Phase 1 (#426)

DB migration and Model changes

eap committed:
Add start button to restart stopped clusters. (#457)

Add a button that will start any cluster whose current state is "Stopped". Rather than adding a dedicated start/stop button, the new button replaces the "Connect" button for clusters in the "Stopped" state.

akarukappadath committed:
fixing PR

akarukappadath committed:
fixing PR

akarukappadath committed:
fixed merge conflicts

rtitle commented:
I think something went wrong with your rebase, because this PR is showing WAY too many commits. :) Not quite sure how to fix it... can you explain how you did the rebase?

rtitle commented:
This can be closed in favor of https://github.com/DataBiosphere/leonardo/pull/473
